### ISSUE TITLE
Extend bounds declaration checking to understand equality.

### ISSUE THREAD

**Issue by dtarditi** [Microsoft/checkedc-clang#474]
The compiler statically checks that bounds declarations are valid.  Bounds declarations must be implied by existing bounds declaration and other information in the environment.    Currently, the compiler doesn't understand that assignment and initialization implies facts about equality of expressions that can be used during checking of bounds declarations.   

This causes the compiler to issue spurious warnings, and is why I've left checking of bounds declarations turned off for unchecked scopes.

This work item is to extend the compiler to collect quality information for individual assignment statements (or declarations with initializers) and use it during checking of bounds declarations.  For example, given
```
array_ptr<int> b : count(5) = ...
array_ptr<int> c : count(5) = b;
```
we are trying to prove for the declaration of `c` that the bounds of `b` imply 
```
bounds(c, c + 5)
```
is true.   The bounds of b is `bounds(b, b + 5)`.    If we know that `c == b`, it is simple to conclude that `bounds(b, b + 5)` implies `bounds(c, c + 5)` is valid.

It is future work to extend this checking to gather and use equality facts across statements/declarations.

**Comment by dtarditi** [comment]
Work is complete.
